pkg/util/validation/exporter/exporter.go

```diff
@@ -180,79 +162,132 @@ func NewOverridesExporter(
+			if tagValue != "" {
+				fieldsByTag[tagValue] = field
+				metricNames = append(metricNames, tagValue)
+			}
+		}
 	}
 
-	// Alertmanager limits
-	if enabledMetrics.IsAllowed(notificationRateLimit) {
-		exportedMetrics = append(exportedMetrics, ExportedMetric{notificationRateLimit, func(limits *validation.Limits) float64 { return limits.NotificationRateLimit }})
+	return &LimitsFieldRegistry{
+		limitsType:         limitsType,
+		fieldsByYamlTag:    fieldsByTag,
+		allowedMetricNames: metricNames,
+		defaultLimitsValue: defaultLimitsValue,
 	}
-	if enabledMetrics.IsAllowed(alertmanagerMaxDispatcherAggregationGroups) {
-		exportedMetrics = append(exportedMetrics, ExportedMetric{alertmanagerMaxDispatcherAggregationGroups, func(limits *validation.Limits) float64 {
-			return float64(limits.AlertmanagerMaxDispatcherAggregationGroups)
-		}})
+}
+
+// GetField returns the struct field for a given yaml tag name
+func (r *LimitsFieldRegistry) GetField(yamlTag string) (reflect.StructField, bool) {
+	field, found := r.fieldsByYamlTag[yamlTag]
+	return field, found
+}
+
+// ValidateMetricName validates that a metric name exists and can be converted to float64
+func (r *LimitsFieldRegistry) ValidateMetricName(metricName string) error {
+	field, found := r.fieldsByYamlTag[metricName]
+	if !found {
+		return fmt.Errorf("enabled-metrics: unknown metric name '%s' - must match a yaml tag in the Limits struct", metricName)
 	}
-	if enabledMetrics.IsAllowed(alertmanagerMaxAlertsCount) {
-		exportedMetrics = append(exportedMetrics, ExportedMetric{alertmanagerMaxAlertsCount, func(limits *validation.Limits) float64 { return float64(limits.AlertmanagerMaxAlertsCount) }})
+
+	// Attempt to convert the field type to float64 to ensure it's supported
+	fieldValue := r.defaultLimitsValue.FieldByName(field.Name)
+	if _, err := convertToFloat64(fieldValue); err != nil {
+		return fmt.Errorf("enabled-metrics: metric '%s' has unsupported type %s - %v", metricName, fieldValue.Type(), err)
 	}
-	if enabledMetrics.IsAllowed(alertmanagerMaxAlertsSizeBytes) {
-		exportedMetrics = append(exportedMetrics, ExportedMetric{alertmanagerMaxAlertsSizeBytes, func(limits *validation.Limits) float64 { return float64(limits.AlertmanagerMaxAlertsSizeBytes) }})
+
+	return nil
+}
+
+// GetAllowedMetricNames returns all available metric names
+func (r *LimitsFieldRegistry) GetAllowedMetricNames() []string {
+	return r.allowedMetricNames
+}
+
+func setupExportedMetrics(enabledMetrics *util.AllowList, extraMetrics []ExportedMetric) []ExportedMetric {
+	var (
+		exportedMetrics []ExportedMetric
+		fieldRegistry   = newLimitsFieldRegistry()
+	)
+
+	// Get all possible metric names from global registry and check if each is enabled
+	for _, metricName := range fieldRegistry.GetAllowedMetricNames() {
+		if enabledMetrics.IsAllowed(metricName) {
+			field, found := fieldRegistry.GetField(metricName)
+			if !found {
+				panic("couldn't find fields that fields registry returned, this shouldn't happen")
```

Comment:
Can we include the field name that triggered the error here?

Reply:
```suggestion
				panic(fmt.Sprintf("couldn't find fields that the fields registry returned, this shouldn't happen (metric name: %s)", metricName))
```